src/extension/chatSessions/vscode-node/copilotCloudSessionsProvider.ts

```diff
@@ -173,25 +174,70 @@ export class CopilotCloudSessionsProvider extends Disposable implements vscode.C
 		@IInstantiationService instantiationService: IInstantiationService,
 		@IGithubRepositoryService private readonly _githubRepositoryService: IGithubRepositoryService,
 		@IChatDelegationSummaryService private readonly _chatDelegationSummaryService: IChatDelegationSummaryService,
+		@IExperimentationService private readonly _experimentationService: IExperimentationService,
 	) {
 		super();
-		const interval = setInterval(async () => {
-			const repoId = await getRepoId(this._gitService);
-			// TODO: handle no auth token case more gracefully
-			if (!this._authenticationService.permissiveGitHubSession) {
-				return;
-			}
-			const sessions = await this._octoKitService.getAllOpenSessions(repoId ? `${repoId.org}/${repoId.repo}` : undefined);
-			if (this.cachedSessionsSize !== sessions.length) {
-				this.refresh();
+
+		// Background refresh
+		getRepoId(this._gitService).then(async repoId => {
+			const telemetryObj: {
+				intervalMs?: number;
+				hasHistoricalSessions?: boolean;
+				isEmptyWindow: boolean;
+			} = {
+				isEmptyWindow: !vscode.workspace.workspaceFolders || vscode.workspace.workspaceFolders.length === 0
+			};
+			if (repoId) {
+				const sessions = await this._octoKitService.getAllSessions(`${repoId.org}/${repoId.repo}`, false);
+				const hasHistoricalSessions = sessions.length > 0;
+				const intervalMs = this.getRefreshIntervalTime(hasHistoricalSessions);
+				telemetryObj.intervalMs = intervalMs;
+				telemetryObj.hasHistoricalSessions = hasHistoricalSessions;
+				const intervalCallback = async () => {
+					// TODO: handle no auth token case more gracefully
+					if (!this._authenticationService.permissiveGitHubSession) {
+						return;
+					}
+					const sessions = await this._octoKitService.getAllSessions(`${repoId.org}/${repoId.repo}`);
+					if (this.cachedSessionsSize !== sessions.length) {
+						this.refresh();
+					}
+				};
+				let interval = setInterval(intervalCallback, intervalMs);
+				this._register(toDisposable(() => clearInterval(interval)));
+				this._register(vscode.window.onDidChangeWindowState((e) => {
+					if (!e.active) {
+						clearInterval(interval);
+					} else if (!interval) {
+						interval = setInterval(intervalCallback, intervalMs);
+						this._register(toDisposable(() => clearInterval(interval)));
+					}
+				}));
```

Comment:
If you want to do the posh experience I mentioned:


```suggestion
				let lastRefreshedAt = 0;
				const scheduler = this._register(new RunOnceScheduler(() => {
					lastRefreshedAt = Date.now();
					this.refresh();
				}, intervalMs));
				scheduler.schedule();

				this._register(vscode.window.onDidChangeWindowState((e) => {
					if (!e.active) {
						scheduler.cancel();
					} else if (!scheduler.isScheduled()) {
						scheduler.schedule(Math.max(0, intervalMs - (Date.now() - lastRefreshedAt)));
					}
				}));
```

something like that so we eagerly fetch/resume the interval once the window is active again

Reply:
This is very cool, didn't know it existed, thanks!